runtime/inspector/mod.rs

```diff
@@ -0,0 +1,722 @@
+  ) -> RefCell<Self> {
+    let new_incoming = new_session_rx
+      .map(move |session_proxy| {
+        WebsocketSession::new(v8_inspector.clone(), session_proxy)
+      })
+      .boxed_local();
+    let self_ = Self {
+      new_incoming,
+      ..Default::default()
+    };
+    RefCell::new(self_)
+  }
+}
+
+impl Default for SessionContainer {
+  fn default() -> Self {
+    Self {
+      new_incoming: stream::empty().boxed_local(),
+      handshake: None,
+      established: FuturesUnordered::new(),
+    }
+  }
+}
+
+struct InspectorWakerInner {
+  poll_state: PollState,
+  task_waker: Option<task::Waker>,
+  parked_thread: Option<thread::Thread>,
+  inspector_ptr: Option<NonNull<JsRuntimeInspector>>,
+  isolate_handle: v8::IsolateHandle,
+}
+
+unsafe impl Send for InspectorWakerInner {}
+
+struct InspectorWaker(Mutex<InspectorWakerInner>);
+
+impl InspectorWaker {
+  fn new(isolate_handle: v8::IsolateHandle) -> Arc<Self> {
+    let inner = InspectorWakerInner {
+      poll_state: PollState::Idle,
+      task_waker: None,
+      parked_thread: None,
+      inspector_ptr: None,
+      isolate_handle,
+    };
+    Arc::new(Self(Mutex::new(inner)))
+  }
+
+  fn update<F, R>(&self, update_fn: F) -> R
+  where
+    F: FnOnce(&mut InspectorWakerInner) -> R,
+  {
+    let mut g = self.0.lock().unwrap();
+    update_fn(&mut g)
+  }
+}
+
+impl task::ArcWake for InspectorWaker {
+  fn wake_by_ref(arc_self: &Arc<Self>) {
+    arc_self.update(|w| {
+      match w.poll_state {
+        PollState::Idle => {
+          // Wake the task, if any, that has polled the Inspector future last.
+          if let Some(waker) = w.task_waker.take() {
+            waker.wake()
+          }
+          // Request an interrupt from the isolate if it's running and there's
+          // not unhandled interrupt request in flight.
+          if let Some(arg) = w
+            .inspector_ptr
+            .take()
+            .map(|ptr| ptr.as_ptr() as *mut c_void)
+          {
+            w.isolate_handle.request_interrupt(handle_interrupt, arg);
+          }
+          extern "C" fn handle_interrupt(
+            _isolate: &mut v8::Isolate,
+            arg: *mut c_void,
+          ) {
+            let inspector = unsafe { &*(arg as *mut JsRuntimeInspector) };
+            let _ = inspector.poll_sessions(None);
+          }
+        }
+        PollState::Parked => {
+          // Unpark the isolate thread.
+          let parked_thread = w.parked_thread.take().unwrap();
+          assert_ne!(parked_thread.id(), thread::current().id());
+          parked_thread.unpark();
+        }
+        _ => {}
+      };
+      w.poll_state = PollState::Woken;
+    });
+  }
+}
+
+/// An inspector session that proxies messages to Websocket connection.
+struct WebsocketSession {
+  v8_channel: v8::inspector::ChannelBase,
+  // TODO(bartlomieju): could probably be v8::UniqueRef instead of UniquePtr
+  v8_session: Rc<RefCell<v8::UniquePtr<v8::inspector::V8InspectorSession>>>,
```

Comment:
Is there really a many-to-one relationship between `WebSocketSession` and `V8InspectorSession`? 
If not, the `Rc` should be unnecessary (and potentially the RefCell too, IDK).

Reply:
At the moment it might seem superfluous, but I want to rewrite `WebsocketSession::receive_from_proxy` function to be implemented as a `Future` instead - then dispatching messages to `v8_session` could be handled with a single helper function which is implemented nicely thanks to wrapping in `Rc<RefCell<>>`